### PR TITLE
Login walkthrough

### PR DESCRIPTION
**Pull request by tjyoung99**
Added the library walk-through's User Authentication feature in order to manage user access to comments page.

### CODE REVIEW

**Comment by tjyoung99** [comment]
https://travisjy-step-2020-v2.ue.r.appspot.com/

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) 
+      throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    String logout = request.getParameter("logout-button");
+    if (logout != null){
+      String logoutUrl = userService.createLogoutURL("/index.html");
+      response.sendRedirect(logoutUrl);
+    }
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:60]: Logout doesn't seem to work quite right on the deployed server? For me it opens a new window with my Google accounts and also doesn't log me out on your site

**Review by kellyktan — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        new Query("UserInfo")
+        .setFilter(new Query.FilterPredicate("id", 
+        Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return "";
+    }
+
+    String nickname = (String) entity.getProperty("nickname");
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java:93]: return directly

File: portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  /** Returns the nickname of the user with id, or null if the user has not set a nickname. */
+  private String getUserNickname(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java:65]: Consider making this a shared function

**Review by suparnabarua — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -79,7 +84,22 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
-    response.sendRedirect("/index.html");
+    response.sendRedirect("/comments.html");
+  }
+
+  private String getUserNickname(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query("UserInfo")
+        .setFilter(new Query.FilterPredicate("id", 
+        Query.FilterOperator.EQUAL, id));
```
[the remark this answers is not in the document]
Comment (tjyoung99) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:95]: so it should look like this?:
Query query = new Query("UserInfo")
        .setFilter(FilterPredicate("id", FilterOperator.EQUAL, id));

and the imports should look like this?:
import com.google.appengine.api.datastore.Query.FilterOperator;
import com.google.appengine.api.datastore.Query.FilterPredicate;

[the remark this answers is not in the document]
Comment (kellyktan) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:95]: Imports: correct.
Code: take a look at my suggested change (you're missing the `new` in front of `FilterPredicate`)

[the remark this answers is not in the document]
Comment (tjyoung99) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:95]: wow I just now realized that I've been looking at the suggested changes thinking that it was my code this whole time.

File: portfolio/src/main/java/com/google/sps/Util/GetNickname.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+
+
+ public class GetNickname {
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/Util/GetNickname.java:12]: nit: Rename to something like NicknameUtil

File: portfolio/src/main/java/com/google/sps/Util/GetNickname.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return "";
+    }
+
+    String nickname = (String) entity.getProperty("nickname");
+    return nickname;
+  }
+}
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/Util/GetNickname.java:28]: newline at end of file

File: portfolio/src/main/java/com/google/sps/Util/GetNickname.java (hunk 5 above)
Comment (kellyktan) [reply portfolio/src/main/java/com/google/sps/Util/GetNickname.java:12]: And package/directory name should be lowercased

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -75,7 +75,9 @@ public void doPost(HttpServletRequest request, HttpServletResponse response)
     long timestamp = System.currentTimeMillis();
 
     UserService userService = UserServiceFactory.getUserService();
-    String user = getUserNickname(userService.getCurrentUser().getUserId());
+    GetNickname getNickname = new GetNickname();
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:78]: For static methods, no need to create class object.  Can just call `GetNickname.getUserNickname()`

**Review by kellyktan — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/Util/GetNickname.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+
+
+ public class GetNickname {
+ /*** Returns the nickname of the user with id, or empty String if the user has not set a nickname.
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/Util/GetNickname.java:13]: nit. Remove the third asterisk

File: portfolio/src/main/java/com/google/sps/Util/GetNickname.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("UserInfo")
+        .setFilter(new FilterPredicate("id", FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      return "";
+    }
+
+    String nickname = (String) entity.getProperty("nickname");
```
Comment (suparnabarua) [inline portfolio/src/main/java/com/google/sps/Util/GetNickname.java:25]: You can just return, without assigning it to the nickname varfiable

**Review by suparnabarua — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/Util/GetNickname.java [hunk 10]
```diff
@@ -1,28 +0,0 @@
...
-import com.google.appengine.api.datastore.DatastoreService;
-import com.google.appengine.api.datastore.DatastoreServiceFactory;
-import com.google.appengine.api.datastore.Entity;
-import com.google.appengine.api.datastore.PreparedQuery;
-import com.google.appengine.api.datastore.Query;
-import com.google.appengine.api.datastore.Query.FilterOperator;
-import com.google.appengine.api.datastore.Query.FilterPredicate;
-
-
- public class GetNickname {
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/Util/GetNickname.java]: Wait why did you delete this? /did you mean to rename it? Not seeing the logic anywhere else (plus the imports in other classes still reference it)

**Review by kellyktan — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/Util/GetNickname.java (hunk 10 above)
Comment (tjyoung99) [reply portfolio/src/main/java/com/google/sps/Util/GetNickname.java]: I've been using commit -am and when I renamed getNickname the file became an untracked file so when I pushed it was excluded. I'll fix and I'll also be changing the name for those references.

File: portfolio/src/main/java/com/google/sps/util/NicknameUtil.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+
+
+ public class NicknameUtil {
+ /** Returns the nickname of the user with id, or empty String if the user has 
+    not set a nickname.
+    */
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/util/NicknameUtil.java:15]: nit: spacing

```suggestion
  /**
   * Returns the nickname of the user with id, or empty String if the user has 
   * not set a nickname.
   */
```

File: portfolio/src/main/java/com/google/sps/util/NicknameUtil.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+
+
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/util/NicknameUtil.java:11]: nit: there's an extra line here.  Also add class javadoc: 
https://docs.google.com/document/d/17zPdB6a-rw5ZBvm-XvZcxv9AyzMYsSHy0B3U1pVA0s8/edit#heading=h.7jt58ds3f8xc

**Review by kellyktan — APPROVED** [review]
LGTM pending final comments

File: portfolio/src/main/java/com/google/sps/util/NicknameUtil.java [hunk 13]
```diff
@@ -8,11 +8,12 @@
 import com.google.appengine.api.datastore.Query.FilterOperator;
 import com.google.appengine.api.datastore.Query.FilterPredicate;
 
-
- public class NicknameUtil {
- /** Returns the nickname of the user with id, or empty String if the user has 
-    not set a nickname.
-    */
+  /** Static utilities for checking user nicknames. */
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/util/NicknameUtil.java:11]:
```suggestion
/** Static utilities for checking user nicknames. */
```

File: portfolio/src/main/java/com/google/sps/util/NicknameUtil.java [hunk 14]
```diff
@@ -8,11 +8,12 @@
...
+public class NicknameUtil {
+ /** 
+  * Returns the nickname of the user with id, or empty String if the user has 
+  * not set a nickname.
+  */
```
Comment (kellyktan) [inline portfolio/src/main/java/com/google/sps/util/NicknameUtil.java:16]: nit: `/**` should left align with the method indentation.
```suggestion
  /** 
   * Returns the nickname of the user with id, or empty String if the user has 
   * not set a nickname.
   */
```

**Review by kellyktan — APPROVED** [review]
LGTM pending my last suggestions.

**Review by suparnabarua — APPROVED** [review]

**Pull request merged**